kiali/kiali
# Trim jaeger project dependencies

Required for https://github.com/kiali/kiali/issues/1372

Kiali queries Jaeger backend using two ways: using the legacy HTTP REST api and the gRPC api_v2.
gRPC protocol is the preferred/default one used in upstream releases but we keep the HTTP legacy one as the gRPC protocol reported problems when used in combination of sidecar containers like OAuth.

As part of a technical debt, Kiali was pulling the whole jaeger project as a dependency in the go.mod:
![image](https://user-images.githubusercontent.com/1662329/134492250-afc191de-bdaf-4c1f-9c5b-dcdf0f2e146c.png)

The main goal of this PR is to remove that dependency and use gRPC using the .proto files defined in Jaeger but without importing any dependency (as it's recommended by Jaeger project).

This is necessary to pull other gRPC dependencies, this work is part of the epic to incorporate istio.io/client-go model into Kiali.

So, in this work you will see:
- Remove of the jaeger dependencies.
- Bring the {model,query}.proto files.
- Update the {model,query}.proto files to remove the "gogo" dependencies to have a clean protobuffer model.
- Bring and clone the minimal dependencies of the jaeger/model to manipulate TraceID and SpanID.
- Bring and clone the minimal dependencies of the jaeger/model to keep the legacy HTTP model and the JSON converted used in the UI layer.

It is being discussed with jaeger that some of the jaeger/model files should be moved to the jaeger-lib to use that dependency with no conflicts, but it's not yet there so cloning a set of minimal classes look reasonable rather than pull the whole project (more info in the main discussion https://github.com/kiali/kiali/issues/1372#issuecomment-923089928 with links to specific issues in jaeger and istio).

In short, with this refactoring, the expectation is that everything works just fine and dependencies are clean as previous step to bring other gRPCs clients.

For QE/Testing:
- Test Kiali with an example like travel-agency that have traces.
- Use the HTTP protocol (point to your backend, but as an example):
![image](https://user-images.githubusercontent.com/1662329/134493311-637f2c2c-33aa-49ce-ace5-b45cca0a629a.png)
- Use the gRPC protocol (point to your backend, but as an example):
![image](https://user-images.githubusercontent.com/1662329/134493367-8eed8b98-8bb2-47e9-94b2-7f005895ae84.png)

Check that there is no regression in tracing features (graph, trace tabs, correlation with metrics and logs, trace details).
Both HTTP and gRPC works well.
